Dewikification: Migrate site content to markdown.

kosayoda opened the pull request:
This PR migrates pages under https://pythondiscord.com/pages/ to the site repo as part of dewikification.

### How to Review
1. Run the branch locally.
2. Compare each page with the corresponding page on the current website.
3. Don't suggest grammar/phrasing changes for any existing page content, however new content like category and page descriptions can be brought up for improvement.
4. Check if links function as intended, and that any images used are local static images and not hotlinked to another website.
5. Get a thank you from me for taking the time to review.

**Notes**
- Certain files have trailing spaces; If the spaces are 2-wide, this is intentional for proper spacing on the displayed markdown. 1-wide trailing spaces can be cleaned up.

**Fixes**
- 10e1727c703d3cb91321e9ef51f54a0117efed19 Removed duplicates from the `Sub-Articles` button in category-page hybrids.

kosayoda committed:
Migrate root files.

kosayoda committed:
Use the more consistent highlight.

We don't need ReasonML support, and the reasonable theme has a really
ugly neon parameter color.

kosayoda committed:
Migrate contributing pages.

Some guides were updated to fit current PyCharm UI and to keep up with
Seasonalbot -> Sir-Lancebot changes.

kosayoda committed:
Add contributing guides assets.

kosayoda committed:
Avoid duplicate entries for page and category.

We have to add another check since `subarticles` is actually a list of
dictionaries and not a dictionary, so key overriding does not work.

We want the categories to take precedence over the pages in case of
future feature expansion involving showing sub-sub-articles.

kosayoda committed:
Migrate Python guides.

kosayoda committed:
Add Python guide assets.

kosayoda committed:
Migrate Python Discord guides.

kosayoda committed:
Add Python Discord guides assets.

kosayoda committed:
Add miscellaneous pages and fixes.

kosayoda committed:
Add missed Game Jam 2020 subpages.

kosayoda committed:
Remove duplicate images.

MarkKoz commented on pydis_site/apps/content/resources/server-info/_info.yml:
```diff
@@ -0,0 +1,3 @@
+title: Server Information
+description: Information on roles, tooling and infrastructure at Python Discord.
```
```suggestion
description: Information on roles, tooling, and infrastructure at Python Discord.
```

kosayoda replied:
Addressed in 5c44a33.

MarkKoz commented on pydis_site/apps/content/resources/guides/python-guides/mutability.md:
```diff
@@ -0,0 +1,52 @@
+---
+title: Mutability and Immutability in Python
+description: "Mutable and immutable data types: What they are and how they work."
```
```suggestion
description: "Mutable and immutable data types: what they are and how they work."
```

kosayoda replied:
Addressed in 5c44a33.

MarkKoz commented on pydis_site/apps/content/resources/security-notice.md:
```diff
@@ -0,0 +1,37 @@
+---
+title: Security Notice
+description: How vulnerabilities in our projects should be reported.
+icon: fas fa-lock
```
The lock icon for the security page may be misleading - my first thought was that it was an inaccesible page. Do you think a shield icon would be better?

kosayoda replied:
Addressed in 5c44a33.

MarkKoz commented on pydis_site/apps/content/resources/guides/python-guides/mutability.md:
```diff
@@ -0,0 +1,52 @@
+![s refers to the string "hello"](/static/images/content/mutability/s_refers_hello.webp)
+
+When you call `s.upper()`, a new string, which contains the characters `'HELLO'`, gets created.
+![s.upper creates "HELLO"](/static/images/content/mutability/s_upper_creates_HELLO.webp)
```
The images in this page are misaligned.
![Screenshot_2021-04-17 Python Discord - Mutability and Immutability in Python](https://user-images.githubusercontent.com/1515135/115124443-1fe5a380-9f77-11eb-8c00-e36add6546a1.png)

kosayoda replied:
Addressed in a782abd.

MarkKoz commented on pydis_site/apps/content/resources/server-info/roles.md:
```diff
@@ -0,0 +1,131 @@
+
+# Basic Roles
+
+### <span class="fas fa-circle" style="color:#6e6e6e"></span> Announcements
```
I like the change to use circles next to role names rather than colouring the text. This is consistent with what Discord did in their client too.

MarkKoz commented on pydis_site/apps/content/resources/code-of-conduct.md:
```diff
@@ -0,0 +1,100 @@
+
+---
+
+# Code of Conduct Policies
```
Was there a discussion about merging the CoC pages into one? Since the content is important, we should be mindful of how a longer page may detract readers. I'm not necessarily for or against it; I just want to make sure some collective thought was given to the decision.

kosayoda replied:
I merged them initially because the page had a custom sidebar, I'll bring it up with the admin team.

MarkKoz commented on pydis_site/apps/content/resources/guides/pydis-guides/helping-others.md:
```diff
@@ -0,0 +1,139 @@
+---
+title: Helping Others
+description: The staff take on how to help others in our community.
```
```suggestion
description: The staff's take on how to help others in our community.
```

kosayoda replied:
Addressed in 5c44a33.

MarkKoz commented on pydis_site/apps/content/resources/guides/pydis-guides/contributing/forking-repository.md:
```diff
@@ -0,0 +1,18 @@
+![Github Fork to User](/static/images/content/contributing/fork_user.webp)
+3. Later, you will need the Git URL of your forked repository in order to clone it.
+In your newly forked repository, copy the Git URL by clicking the green `Code` button, then click the Copy Link button.
+![Github Fork Clone URL](/static/images/content/contributing/fork_clone.webp)
```
This would benefit from some extra padding between the text and the images.

MarkKoz replied:
This actually applies to all images in guides, but I don't consider it too important.

kosayoda replied:
I found that it only applies to images in lists, as in the other cases it would be separated by paragraph padding.

One solution was to add even more `<br>` tags (we can't put a newline in the markdown otherwise the image would extend past the list item's left margin), but I've decided to use CSS to avoid having article writers think about it.

MarkKoz commented on pydis_site/apps/content/resources/guides/pydis-guides/contributing/cloning-repository.md:
```diff
@@ -0,0 +1,31 @@
+![Create Project in PyCharm](/static/images/content/contributing/pycharm_create_project.webp)
+2. Enter the URL of your forked repository.
+3. Change the directory if you desire and click `Clone`.
+![Clone Git Project in Pycharm](/static/images/content/contributing/pycharm_checkout.webp)
```
Similar issue of images being on the same line as the text rather than on a new line.

MarkKoz commented on pydis_site/apps/content/resources/guides/pydis-guides/contributing/cloning-repository.md:
```diff
@@ -0,0 +1,31 @@
+icon: fab fa-github
+---
+
+> **Note:** The process varies depending on your choice of code editor / IDE, so refer to one of the following guides:
+
+- [Cloning with PyCharm](#cloning-with-pycharm)
+- [Cloning with the command line](#cloning-with-the-command-line)
```
Would this be better as a table of contents or would that make the note confusing?

MarkKoz commented on pydis_site/apps/content/resources/guides/pydis-guides/contributing/installing-project-dependencies.md:
```diff
@@ -0,0 +1,38 @@
+icon: fab fa-python
+---
+
+> **Note:** The process varies depending on your choice of code editor / IDE, so refer to one of the following guides:
+
+- [Installing dependencies with PyCharm](#installing-dependencies-with-pycharm)
+- [Installing dependencies with the command line](#installing-dependencies-with-the-command-line)
```
Would this be better as a table of contents or would that make the note confusing?

MarkKoz commented on pydis_site/apps/content/resources/guides/pydis-guides/contributing/sir-lancebot.md:
```diff
@@ -0,0 +1,119 @@
+
+Notably, version control of our projects is done using Git and Github. It can be intimidating at first, so feel free to ask for any help in the server.
+
+[**Click here to see the basic Git workflow when contributing to one of our projects.**](./working-with-git/)
```
This link does not resolve.

kosayoda replied:
Addressed in 7a1d1ae.

MarkKoz commented on pydis_site/apps/content/resources/guides/pydis-guides/contributing/sir-lancebot.md:
```diff
@@ -0,0 +1,119 @@
+| `CHANNEL_DEVLOG` | ID of the `#dev-log` channel |
+| `CHANNEL_COMMUNITY_BOT_COMMANDS` | ID of the `#sir-lancebot-commands` channel |
+
+[**Full environment variable reference for this project.**](./sir-lancebot-env-var-reference)
```
This link does not resolve.

kosayoda replied:
Addressed in 7a1d1ae.

MarkKoz commented on pydis_site/apps/content/resources/guides/pydis-guides/contributing/docker.md:
```diff
@@ -0,0 +1,119 @@
+> NOTE: Solution requires Windows user credentials for an account that has administrative privileges.
+
+1. Right-click the Docker icon in the Windows system tray, and choose "Settings" from the context menu.
+![Docker Settings](/static/images/content/contributing/docker_settings.webp)
```
This image has alignment issues like the other mentioned cases.

kosayoda replied:
Addressed in a782abd.

MarkKoz commented on pydis_site/apps/content/resources/guides/pydis-guides/contributing/working-with-git/pycharm.md:
```diff
@@ -0,0 +1,75 @@
+> Before making new branches, be sure to checkout the `main` branch and ensure it's up to date.
+
+1. In the bottom right corner, click on `main` and then click `New Branch`.    
+![PyCharm New Branch](/static/images/content/contributing/pycharm_branch.webp)
```
This image is especially blurry/low res and appears to be scaled up. It seems like that's cause the original image was already pretty bad and the compression artefacts you've added made aggravated it.

kosayoda replied:
This is solved with the move back to PNG.

MarkKoz commented on pydis_site/apps/content/resources/guides/pydis-guides/contributing/working-with-git/cli.md:
```diff
@@ -0,0 +1,131 @@
+* [Discarding Changes](#discarding-changes)
+* [Committing Changes](#committing-changes)
+* [Pushing Commits](#pushing-changes)
+* [Pulling Changes](#pulling-changes)
```
I think a ToC would be better. I don't find the distinction between frequently particularly useful. Same applies to the PyCharm guide.

kosayoda replied:
Addressed in 0697d89.

MarkKoz reviewed (changes requested):
It'd be helpful for the following pages to have a table of contents:

* _FAQ_
* _Code Reviews: A Primer_
* _Function Parameters and Arguments in Python_
* _Helping Others_
* _Help Channels_
* _Asking Good Questions_
* _Working with Docker & Docker Compose_
* _Preparing Your Hosts file_

Thoughts?

---

None of the "more" links in the navbar work. Will that be addressed later?

---

The compression artefacts on many images are fairly noticeable. Can you use higher-quality images?

kosayoda committed:
Improve markdown whitespace.

Replace double trailing whitespaces with <br> tags to prevent certain
editors from automatically stripping the necessary space.

Remove any extra single whitespace.

kosayoda committed:
Replace WEBP images with their PNG counterparts.

kosayoda committed:
Update references to static images.

kosayoda committed:
Add TOC to certain pages.

kosayoda committed:
Implement miscellaneous review changes.

Punctuation, icon usage...

kosayoda committed:
Fix unresolving links.

kosayoda committed:
Increase padding between text and images in lists.

This has a caveat in that there may be additional padding if the first
thing in a list item is an image, but it is very unlikely in practice,
so it's not handled for now.

MarkKoz commented on pydis_site/static/css/content/page.css:
```diff
@@ -68,3 +68,12 @@ ul.menu-list.toc {
+ * This is used instead of a newline between the image and the text in markdown
+ * so the image remains left-aligned with the list item.
+ */
+li img {
```
Why not do this for both ordered and unordered lists?

kosayoda replied:
Ordered and unordered lists both use the `<li>` tag for list items, which is what I'm styling here, so it works for both.

Xithrius reviewed (approved):
I don't know html and css that much, but it all looks good to me.

whispersofthedawn reviewed (approved):
The HTML and CSS looks good to me.
